File: 105-从中序与后序遍历序列构建二叉树.py
class Solution:
    def build_tree(self, inorder, postorder, a, b, c, d):
        if a == b:
            return None

        i = inorder.index(postorder[d - 1])

        root = TreeNode(postorder[d - 1])

        root.left = self.build_tree(inorder, postorder, a, i, c, c + i - a)
        root.right = self.build_tree(inorder, postorder, i + 1, b, c + i - a, d - 1)

        return root

    def buildTree(self, inorder: List[int], postorder: List[int]) -> TreeNode:
        return self.build_tree(inorder, postorder, 0, len(inorder), 0, len(postorder))

File: 105-从中序与后序遍历序列构建二叉树.cpp
class Solution {
public:
    TreeNode* build_tree(vector<int>& inorder, vector<int>& postorder, int a, int b, int c, int d){
        if(a == b)
            return NULL;

        int i = a - 1;
        while(inorder[++i] != postorder[d - 1]);

        TreeNode* root = new TreeNode(postorder[d - 1]);

        root->left = build_tree(inorder, postorder, a, i, c, c + i - a);
        root->right = build_tree(inorder, postorder, i + 1, b, c + i - a, d - 1);

        return root;
    }
    
    TreeNode* buildTree(vector<int>& inorder, vector<int>& postorder) {
        return build_tree(inorder, postorder, 0, inorder.size(), 0, postorder.size());
    }
};
